### PR TITLE
Add upstream patch to remove internal types in favor of unstructured

### PR DESCRIPTION
**Pull request by elmiko**
This pull request brings in the upstream changes encapsulated in https://github.com/kubernetes/autoscaler/pull/3312

it also adds a commit which reintroduces the openshift-specific changes that were removed from the upstream during the conversion. the additions that are carried on top of the upstream changes are:

* Adds missing imports to the clusterapi_nodegroup.go file.
* Adds helper functions to get labels, taints, and the resource information associated with scaling from zero.
* Fixes the unit tests to use the unstructured package helper functions for inspecting the objects.
* Removes the clusterapi_converters_test.go file as this is no longer needed. All objects are now unstructured by default, there is no need for the converters.
* Removes the fallback from spec.replicas to status.replicas for scalable resources. This change can be safely made because the replicas now come from a scale subresource which cannot return `nil` for its value. This is noted in code comments.
* Adds unit tests for unstructured annotations to ensure that the cpu, memory, gpu, and max pods information is properly parsed.
* Adds unit tests to ensure that the logic for scaling from zero is properly observed.

### CODE REVIEW

**Comment by elmiko** [comment]
there is a permission issue that needs to be addressed before this patch can be merged, see https://github.com/openshift/cluster-autoscaler-operator/pull/177

it appears that the permission issue may have been hiding an error related to missing node selector information, i am continuing to add to this patch.
/hold

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go
Comment (JoelSpeed) [inline cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go:160]: I think this should be this?
```suggestion
	labels, found, err := unstructured.NestedStringMap(r.unstructured.Object, "spec", "template", "spec", "metadata", "labels")
```

Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go:160]: lol, funny. i was just changing this XD

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured_test.go
Comment (JoelSpeed) [inline cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured_test.go:30]: Why do we have to cast one but not the other?

**Review by JoelSpeed — COMMENTED** [review]
Looks good! I think we need to fix the `FailureMessage`/`ErrorMessage` discrepancy and I've got some comments through the tests, nothing too major I don't think though, PTAL

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go
Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:433]: i will have to double check on that

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go
Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:514]: definitely

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go
Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:517]: i _think_ that should work, i have a feeling i was following the changeset here but i can certainly move this logic into the Replicas() call.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go
Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go:98]: yeah, i think the logic for the fallback should be moved here. i don't remember why i had it in the other place but not here, it seemed to make sense at the time.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go
Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go:458]: this piece of code came from the upstream change. i think if we improve it we should do that in the upstream to reduce our carries around this.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go
Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go:473]: yeah, makes sense. i'll ask Jason about his inspiration for these calls.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go
Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go:1598]: i don't _think_ so, but this is part of the upstream changeset.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup_test.go
Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup_test.go:949]: ++, i'll do a little more digging here

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured_test.go
Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured_test.go:30]: this came from the upstream change, i'll see if i can figure it out

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go
Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go:433]: looked into this and `errorMessage` is what we use, i am changing this code

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup_test.go
Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup_test.go:949]: changed to use `ErrorMessage`

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go
Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go:458]: i tried this with `PollImmediate` and things seemed to work well, so i made this patch upstream https://github.com/kubernetes/autoscaler/pull/3652

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured_test.go
Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured_test.go:30]: the main difference is that `updatedReplicas` is used more directly with the k8s api, whereas `initialReplicas` is used with internal test functions.

`initialReplicas` is _mostly_ used for passing the node count into the functions that create the test MachineSets and MachineDeployments. that said, there are 2 places where it is cast as an `int32` to compare with values from the api.

given the way the functions are written, we could clean this up in upstream and make them all `int32`s, but it is solidly in the code cleanliness refactor territory(and even that is debatable).

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go
Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go:473]: see https://github.com/openshift/kubernetes-autoscaler/pull/177#discussion_r514558189

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go
Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go]: i have reverted this code back to something like the original. since we are using the scale resource we should never reach a state where we have the nil pointer here.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go
Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go:98]: see https://github.com/openshift/kubernetes-autoscaler/pull/177#discussion_r515349316

**Comment by elmiko** [comment]
since this change removes the logic we had implemented for a `spec.replicas` backup to `status.replicas` in the event of a `nil` pointer, i am dropping this link[0] to the scale subresource replicas code. using this subresource it should not be possible to get a `nil` value for replicas, and at the worst we would get the defaulted value of `1`.

**Comment by elmiko** [comment]
this patch must go before #179

**Comment by elmiko** [comment]
rebased

**Comment by elmiko** [comment]
this need to merge before #179

**Review by Danil-Grigorev — APPROVED** [review]
/lgtm

**Pull request merged**
